alpenhorn/update.py

```diff
@@ -0,0 +1,466 @@
+
+            # First try bbcp which is a fast multistream transfer tool. bbcp can
+            # calculate the md5 hash as it goes, so we'll do that to save doing
+            # it at the end.
+            if util.command_available('bbcp'):
+                cmd = 'bbcp -f -z --port 4200 -W 4M -s 16 -o -E md5= %s %s' % (from_path, to_dir)
+                ret, stdout, stderr = util.run_command(cmd)
+
+                # Attempt to parse STDERR for the md5 hash
+                if ret == 0:
+                    mo = re.search('md5 ([a-f0-9]{32})', stderr)
+                    if mo is None:
+                        log.error('BBCP transfer has gone awry. STDOUT: %s\n STDERR: %s' % (stdout, stderr))
+                        ret = -1
+                    md5sum = mo.group(1)
+                else:
+                    md5sum = None
+
+            # Next try rsync over ssh. We need to explicitly calculate the md5
+            # hash after the fact
+            elif util.command_available('rsync'):
+                cmd = ("rsync -z%s --rsync-path=\"ionice -c2 -n4 rsync\" -e \"ssh -q\" %s %s" %
+                       (RSYNC_FLAG, from_path, to_dir))
+                ret, stdout, stderr = util.run_command(cmd)
+
+                md5sum = util.md5sum_file(to_file) if ret == 0 else None
+
+            # If we get here then we have no idea how to transfer the file...
+            else:
+                log.warn("No commands available to complete this transfer.")
+                ret = -1
+
+        # Okay, great we're just doing a local transfer.
+        else:
+
+            # First try to just hard link the file. This will only work if we
+            # are on the same filesystem. As there's no actual copying it's
+            # probably unecessary to calculate the md5 check sum, so we'll just
+            # fake it.
+            try:
+                link_path = os.path.join(node.root, req.file.acq.name, req.file.name)
+
+                # Check explicitly if link already exists as this and
+                # being unable to link will both raise OSError and get
+                # confused.
+                if os.path.exists(link_path):
+                    log.error('File %s already exists. Clean up manually.' % link_path)
+                    ret = -1
+                else:
+                    os.link(from_path, link_path)
+                    ret = 0
+                    md5sum = req.file.md5sum  # As we're linking the md5sum can't change. Skip the check here...
+
+            # If we couldn't just link the file, try copying it with rsync.
+            except OSError:
+                if util.command_available('rsync'):
+                    cmd = "rsync -%s %s %s" % (RSYNC_FLAG, from_path, to_dir)
+                    ret, stdout, stderr = util.run_command(cmd)
+
+                    md5sum = util.md5sum_file(to_file) if ret == 0 else None
+                else:
+                    log.warn("No commands available to complete this transfer.")
```

Comment:
Sure. I guess there's a few advantages to using rsync, mostly that it'll resume a failed transfer which could be useful. Let me look at changing it, as just using an internal stdlib would be nicer. Again, I'll add it as a todo list